### PR TITLE
[MIG][10.0] product_multi_image

### PR DESCRIPTION
**Pull request by lasley**
This is WIP & served to rule out old API wire crossing as the 9.0 upgrade issue in #199. We're getting the same effect here, so that's not the case.

I'll fix this PR up once the issue is figured out in the 9.0 version. 

Depends:
- [x] #199 
- [x] OCA/server-tools#600
- [x] OCA/server-tools#629
- [x] OCA/server-tools#677
- [x] OCA/server-tools#709

### CODE REVIEW

**Comment by pedrobaeza** [comment]
You can rebase now

**Comment by lasley** [comment]
Rebased. I don't think it made any difference though? I'm still fixing the variant removal issue noted in #199

**Comment by lasley** [comment]
This is solved in OCA/server-tools#629 & should be ready for review after merge

File: oca_dependencies.txt
Comment (pedrobaeza) [inline oca_dependencies.txt]: Change this to regular repo

**Comment by lasley** [comment]
Restarted build now that base fix is merged. Hopefully we go 🍏 !

**Comment by lasley** [comment]
Oh crap this is still pending OCA/server-tools#677 - updating issue description

**Comment by lasley** [comment]
Alright rebased now the OCA/server-tools#677 is merged. We should hopefully be 🍏  here now.

@pedrobaeza @yajo - you guys mind if I squash your commits into changesets by author?

**Comment by yajo** [comment]
No problem on my side. Still ❌ though.

**Comment by pedrobaeza** [comment]
Yes, please squash, and check Travis status.

**Comment by lasley** [comment]
Ok so it seems the registry does not operate the same way that we expect it to in the hooks. I totally remember now why I went down the path of making an environment in my server-tools PR. It's wanting a `BaseModel` instance passed into the search call, but the model being returned from the registry here is a [`MetaModel`](https://travis-ci.org/LasLabs/product-attribute/jobs/193939662#L834).

I'm submitting a new PR for the base & linked this PR against it. We will use this one to determine that my fix actually worked. Don't merge this PR until I switch back 😉 

_Edit - the MetaModel linked to Travis, which overwrote when I rebuilt - sorry!_

**Comment by yajo** [comment]
OK so now it's 💚, but... are we allowed to merge it like this?

**Comment by pedrobaeza** [comment]
What are you referring to?

File: oca_dependencies.txt
Comment (yajo) [inline oca_dependencies.txt]: @pedrobaeza I'm talking about this.

Comment (pedrobaeza) [reply oca_dependencies.txt]: Ah, OK. No, then we should wait for the merge on the server-tools repo.

Comment (lasley) [reply oca_dependencies.txt]: Yeah please don't merge yet. I'm just using this as proof that my patch in server-tools actually works this time. I didn't do it last time & now we're in this situation. Sorry, I should have been more clear in my comment.

**Comment by lasley** [comment]
Alright at long last, these tests should pass! 

Well, except for the fact that they won't pass due to OCA/maintainer-quality-tools#427 😆

**Comment by lasley** [comment]
Hrm no idea how tests are actually passing with the missing headers, but I'll take it!

**Comment by lasley** [comment]
Without futher ado, I present a 🍏  build - using the production server-tools! Ready for review 🎸 

Btw nobody rebuild Travis or it'll go 🔴  on unrelated issue in odoo/odoo. Delicate tip-toe, this passing CI is.

**Review by pedrobaeza — APPROVED** [review]
Please include also https://github.com/OCA/product-attribute/pull/229

**Comment by lasley** [comment]
Added

**Comment by lasley** [comment]
@tedsalmon please review

**Comment by angelmoya** [comment]
Checked on local environment and works nice!
:+1:

**Comment by pedrobaeza** [comment]
@lasley have you checked if in uninstallation product images are restored?

**Comment by angelmoya** [comment]
@pedrobaeza @lasley Just I try to uninstall and it fail

  File "/opt/odoo/sources/odoo/odoo/modules/loading.py", line 389, in load_modules
    getattr(py_module, uninstall_hook)(cr, registry)
AttributeError: 'module' object has no attribute 'uninstall_hook'

**Comment by pedrobaeza** [comment]
hehe, that why I was asking. It seems anyway something easy to fix.

**Comment by lasley** [comment]
Hahahah this is the PR that just keeps on giving- I'll fix it up, thanks for the catch guys!

**Comment by lasley** [comment]
@pedrobaeza @angelmoya - I added the import to the uninstall hook into the init file, which cleared up the uninstall issue in local.

File: product_multi_image/hooks.py
Comment (pedrobaeza) [inline product_multi_image/hooks.py:22]: You still need to restore main image for products/templates here, before calling the base_multi_image hook, or you will lose them on uninstallation.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: product_multi_image/hooks.py
Comment (lasley) [reply product_multi_image/hooks.py:22]: Ah yes good call. Any idea how we would go about that with the DB modifications still in place? The fields in the product will still be overridden as related, so any changes made to those will simply update the underlying image record instead of the binary field.

Comment (pedrobaeza) [reply product_multi_image/hooks.py:22]: My best bet would be to create the column in SQL, and copy data to this column also in SQL. ORM wouldn't interfere in that, and when the field definition with the compute dissapears, Odoo will look again to this column for getting the image.

**Comment by thomaspaulb** [comment]
@pedrobaeza Since this was superseded by #359, can you close this?

**Comment by pedrobaeza** [comment]
Yeah, closing.

**Pull request closed without merging**
